### ISSUE TITLE
Re-reading config parameters and applying them without server restart

### ISSUE THREAD

**Issue by pavel-zotov** [FirebirdSQL/firebird#7712]
Some server-side config parameters may change often, but this requires server restart or (if parameter is 'per database') disconnect all users from DB.
It will be very useful if server could re-read config and apply changed parameters (if it is possible).
Following parameters can be checked by server (say, once per 5...10 minutes) and be applied with reflecting in rdb$config table:

replication.conf: 
```
   verbose = on | off
   apply_idle_timeout
   apply_error_timeout
```
firebird.conf:
```
    BugcheckAbort
    ExtConnPoolLifeTime /  ExtConnPoolSize
    ParallelWorkers / MaxParallelWorkers
    DSQLCacheSize

    StatementTimeout
    InlineSortThreshold
    TempCacheLimit
    TempBlockSize
    TempTableDirectory
    TempDirectories
```

**Comment by aafemt** [comment]
Do you mean that these changes must be seen during session or it is enough if they apply to new connections only (as now in classic mode)?

**Comment by pavel-zotov** [comment]
I don't know the detals ('mechanics' of how engine works) but ideally most of them must be applied to all existing and new connections.

New value of BugCheckAbort must be seen immediately, without re-connecting anyone.
Changes of ParallelWorkers / MaxParallelWorkers must be seen only for new runs of gbak/sweep.

Of course, ExtConnPool must be cleared when ExtConn* parameters are changed.
The same about DSQLCacheSize.
Changing of TempTableDirectory, TempBlockSize, TempDirectories, StatementTimeout and  InlineSortThreshold must be seen for all subsequent statements of any existing connections.
Most problematic is TempCacheLimit (as it seems to me): its increasing must be allowable, but reducing can be done only when there are no statements which uses external sort.

**Comment by aafemt** [comment]
Theoretically it is possible using ReadDirectoryChanges/inotify and centralized config and may be ASTs for critical changes but "include" clause complicates things.

**Comment by pavel-zotov** [comment]
Why not repeat the same code that reads  .conf-content initially (when server is starting) ? AFAIU, this code without any problems 'dives' in files which are specified in include directive.

**Comment by hvlad** [comment]
The problem not in the reading config values. The problem is what to do with this new values, how to apply it to the working system. Some settings is relatively easy to apply, some - almost impossible. 

I, personally, see no big value in this feature. We already able to change some settings using SQL interface, and config is used to set defaults for them. This is way to go, imho.

**Comment by aafemt** [comment]
Because I/O is time-consuming. Currently config files are read in quite rare points: server start, attach and plugin initialization. Doing it by timer or at every value demand shall degrade performance.

**Comment by pavel-zotov** [comment]
>  Doing it by timer or at every value
ONLY by timer.
About performance affect: what time must be spent to read and parse 4k text file ?.. 30 mc ?

**Comment by aafemt** [comment]
It depends on storage system and its load. Firebird used to be I/O hungry by itself and additional random reads won't make things better.

BTW, the timer is the worst way to implement it because it would generate I/O even when no files has been changed and delay between configuration change and new values to be applied is unpredictable.

**Comment by pavel-zotov** [comment]
> We already able to change some settings using SQL interface
Only ExtConnPool* and StatementTimeout can be changed (I've included them in the list by mistake).
As for me, most desirable are following:
1) from replication.conf: verbose = on | off;
2) from firebird.conf:
   InlineSortThreshold
   TempCacheLimit

**Comment by pavel-zotov** [comment]
> it would generate I/O even when no files has been changed 
Do you mean querying to OS about firebird.conf (and all included parts) attibutes  - size/last access time etc ?
Even if so, one may to change config and explicitly say to engine: "RE-READ it", using somewhat like "alter system reread config;" (yes, one more new command :))

> and delay between configuration change and new values to be applied is unpredictable. 
If time interval was set to some reasonable value (say, 5 minutes), then i see no problem in waiting for 10 minutes to new values be applied.

**Comment by AlexPeshkoff** [comment]
@hvlad support which is realistic to provide is ability to restart server from services, which will force it to re-read all .conf files. That will detach all connections, including sevices. This functionality to be turned off by default and activated only by special firebird.conf parameter. Users who need that are free to set that parameter to true.

**Comment by dyemanov** [comment]
I'm against timers and the like. On Linux it's quite a common practice to use SIGHUP to reload configuration. And of course we could add a session management SQL statement to reload our config. So it's just a matter of suitable implementation. But Vlad has already pointed to more important issue - how to utilize the new values properly. We should start from that side, IMHO.

**Comment by hvlad** [comment]
@AlexPeshkoff every OS have ability to restart service, isn't is ?
@dyemanov, exactly

**Comment by hvlad** [comment]
@AlexPeshkoff btw, the issue subject explicitly state: "without server restart" ;)

**Comment by aafemt** [comment]
On Windows "Restart Service" signal is working exactly as SIGHUP on Linux so there is no problem to use it for reloading config only.

> But Vlad has already pointed to more important issue - how to utilize the new values properly.

We already have centralized config manager which is used by server and (almost) every plugin, right? So we can refresh its storage whenever we like. Those subsystems that use the config do receive needed parameters whenever they like and cache them as long as they are supposed to be stable (up to instant usage w/o caching at all). IMHO something like that.

**Comment by AlexPeshkoff** [comment]
As far as I understood Pavel, he needs a way to avoid OS specific 
restarts of server when running tests. If not I hardly understand what 
for is all that needed.

**Comment by pavel-zotov** [comment]
> he needs a way to avoid OS specific restarts of server when running tests
No, it is not because of tests.
I want (first of all) to avoid restart FB when changing replication verbose to ON/OFF  :-)
Also, changing of InlineSortThreshold and TempCacheLimit 'on the fly' can be very useful feature for productions when we investigate performance problems and find out something related to these parameters (say, extremely heavy sorts).
Finally, changing BugCheckAbort to 1 (on Windows) can help if there was crash(es) but no dump was created.

**Comment by asfernandes** [comment]
@pavel-zotov please do not put your answers inside with what you are quoting, it makes difficult to understand things.

**Comment by pavel-zotov** [comment]
@asfernandes 
yes, I've noted this yesterday: it seems that empty line must be inserted after quoted one.
Sorry for inconvenience, previous messages have been edited/corrected.

**Comment by PizzaProgram** [comment]
I agree with the concept of re-reading the config without restarting the server/service.
Although I see more problems:
 - Checking "modified" time is not safe whenever summer-time (non-UTC) changes under Windows OS. 
 - What happens if there is a bad character or other misconfig in the `.conf` file? _(for example accidentally activating save command or having the text editor an auto-save option...)_

IMHO the solution would be to:
## Triggering the server to try re-read the config files by 1-1 simple select command, like:

`SELECT $RE_READ_CONFIG FROM RDB$DATABASE;` 
`SELECT $RE_READ_REPLICATION FROM RDB$DATABASE;` 

If there is an **error** while reading the new config, an exception will be triggered, which may even explain where the error is.
It would also prevent from accidental changes, and even activation could be planned, when to apply.

There could be a comment/hint/help line about this feature at the top config files too.

**Comment by mrotteveel** [comment]
> `SELECT $RE_READ_CONFIG FROM RDB$DATABASE;`
> `SELECT $RE_READ_REPLICATION FROM RDB$DATABASE;`

I do not like this, this is not a syntax that should be used. Although `$` is not valid as the first character of an unquoted column identifier, this is too close to the normal column syntax. Something that reloads the configuration should be done either with a stored procedure (e.g. `execute procedure RDB$CONFIG_UTILS.RELOAD`), or possibly something similar to the session reset (e.g. `alter config reload`), or maybe even a standalone management statement (e.g. `reload config`).

**Comment by PizzaProgram** [comment]
> Although $ is not valid as the first character of an unquoted column identifier

You are right, that could be a problem, even if the engine and our eyes could easier distinguish from normal command, different APIs may have a pre-parser that can not handle that properly. 
`READ_CONFIG$` would be OK, but I agree:
 - your other idea is even better:

> ... should be done with a stored procedure

## Using procedure style would allow to set proper user-rights too
 _because different users can be added or banned to execute this._
`execute procedure RDB$CONFIG_UTILS.RELOAD;`
`execute procedure RDB$REPLICATION_UTILS.RELOAD;`

By default **only** `SYSDBA` would have proper rights to execute.

Also I would strongly recommend to separate to 2 commands to reload the 2 types of the configs.

**Comment by PizzaProgram** [comment]
PS: just for comparison: PostgreSQL has 3 methods to reload config, including a simple: `select pg_reload_config() ;`
Found out recently from [here ...](https://stackoverflow.com/a/71322705)

Also there are security considerations that should disallow to gain access to a server through operations like this...

**Comment by aafemt** [comment]
Tell me please that these comments were not written by AI...

Though referenced article counts three methods there is actually only two because systemctl just calls pg_ctl.

**Comment by livius2** [comment]
I do not suppose that this should be done by read whole config. 
People will then need to know what values server support to read again, and what not.
Is not better to have it with explicit parameter and name? 
`execute procedure RDB$CONFIG_UTILS.RELOAD('CONFIG_VALUE_NAME'); `
